rfd/XXXX-cloud-agent-upgrades.md

```diff
@@ -0,0 +1,411 @@
+│     │   │                             │
+│     │   │                             │
+│    yes  └──────no──────────► Refresh Schedule
+│     │                                 │
+│     │                                 │
+└─────┴───────────────────────────ok────┘
+```
+
+
+### Kubernetes Model
+
+Kubernetes based agents will be managed by a separate cloud-stable upgrade controller. The upgrade controller will
+be responsible for all teleport agents installed by the helm chart, and will perform rolling
+updates, applying the latest immutable image tag to all agents.
+
+Agents running in kuberentes mode will periodically export their schedules to a kubernetes config object to be
+consumed by the controller.
+
+The k8s controller's specific loop will end up looking like this:
+
+![Upgrade Controller](https://user-images.githubusercontent.com/16366487/219150279-9255d4b3-74fb-48f7-b966-04cb79520e66.png)
+
+The kuberenetes controller will leverage cosign based image signing for an added layer of security.
+
+Our kube agent helm chart will be updated to support enabling all necessary cloud-stable behaviors
+(added controller and any necessary command/env changes) via a single flag. Ex:
+
+```
+$ helm install teleport-kube-agent [...] --set cloudStableUpgrades=yes
+```
+
+
+### Apt/Yum Model
+
+For `apt` and `yum` based installs, we operate under the assumption that teleport is managed by `systemd`. A new
+`cloud` "channel" will be set up, and a new `teleport-cloud-updater` package will be made available on that channel,
+which will depend on the standard teleport package and install additional systemd units.
+
+In order to minimize the risk of users accidentally manually upgrading to an agent version *too new* for the cloud
+control plane, the `cloud` channel will lag a bit behind master.
+
+Note that the behavior of the *repository* "channels" differs from the behavior of the `cloud-stable` TLS endpoint
+"channel".  The repo channels contain all package versions that *have* or *might* become a target of the `cloud-stable`
+channel endpoint. The `cloud-stable` channel endpoint serves the specific target version that agents *aught* to be
+running at the current time.
+
+When a user installs `teleport-cloud-updater`, the main `teleport` package will be brought in as a dependency. The
+`teleport-cloud-updater` package itself will install a `teleport-upgrade` script, and a pair of additional systemd
+timer units: `teleport-upgrade.timer` and `teleport-upgrade-check.timer`. It will also install a systemd override file
+for the main teleport service file adding an `ExecStopPost` command clearing the overrides currently applied to
+`teleport-upgrade.timer` (more on this later).
+
+The `teleport-upgrade` script will be the most trivial of the updater components. It will attempt to install
+the current target version of the `teleport` package as specified by the `cloud-stable/version` endpoint, restarting
+the main teleport service if appropriate.
```

Comment:
Assuming the existing teleport agent is healthy, installing a new version and restarting the service should be considered a risk. Have we considered a model of installing the new version alongside the existing current known-good version, starting the new version and giving it some unit of time to become healthy, else rolling back to the known-good version? This is common among firmware upgrade procedures for devices I have worked on, I half expected to see something like it here, I think it could be very beneficial in avoiding the risk of breaking the install for such a critical access tool.

Reply:
We did consider doing something of this nature (basically a shim/proxy binary that can select the appropriate version to run, and can mark versions as bad to effect a rollback)...

Ultimately, we decided that doing this kind of system *well* was too complex for the timeline we were working with.  We are still interested in doing something like this at some point. For now, we opted to focus on the more modest goal of ensuring that the system couldn't orphan agents after a bad update. Hence the use of a separate updater that continues to function if teleport itself is totally broken, and the general push to keep the system as dumb/simple as possible.

Reply:
Yeah I think it makes sense to cut scope, the updater will still be working so things should be fine as long as we don't ship an update that borks the agent's local state